src/passes/J2CLItableMerging.cpp

```diff
@@ -0,0 +1,373 @@
+        auto itableSize =
+          itableStructNew->type.getHeapType().getStruct().fields.size();
+
+        if (itableSize == 0) {
+          // Itables are empty. Nothing to do.
+          return;
+        }
+
+        // Compute the new size of the vtable.
+        Index newSize = curr->operands.size() + itableSize;
+        // and resize the struct.new operands to accommodate the itable
+        // fields.
+        curr->operands.resize(newSize);
+
+        // Move initialization for the existing vtable fields to their
+        // new position.
+        for (Index i = newSize - 1; i >= itableSize; i--) {
+          curr->operands[i] = curr->operands[i - itableSize];
+        }
+
+        // Add the initialization for the itable fields.
+        for (Index i = 0; i < itableSize; i++) {
+          if (itableFieldInitializers.size() > i) {
+            // The itable was initialized with a struct.new, copy the
+            // initialization values.
+            curr->operands[i] = ExpressionManipulator::copy(
+              itableFieldInitializers[i], *getModule());
+          } else {
+            // The itable was initialized with struct.new_default. So use
+            // null values to initialize the itable fields.
+            Builder builder(*getModule());
+            curr->operands[i] =
+              builder.makeRefNull(itableStructNew->type.getHeapType()
+                                    .getStruct()
+                                    .fields[i]
+                                    .type.getHeapType());
+          }
+        }
+      }
+    };
+
+    Reindexer reindexer(*this);
+    reindexer.run(getPassRunner(), &wasm);
+    reindexer.runOnModuleCode(getPassRunner(), &wasm);
+  }
+
+  // Redirects all itable access by changing `struct.get` of the `itable` field
+  // to `struct.get` on the to `vtable` field.
+  void rerouteItableAccess(Module& wasm) {
+    struct Rerouter : public WalkerPass<PostWalker<Rerouter>> {
+      bool isFunctionParallel() override { return true; }
+
+      J2CLItableMerging& parent;
+
+      Rerouter(J2CLItableMerging& parent) : parent(parent) {}
+
+      std::unique_ptr<Pass> create() override {
+        return std::make_unique<Rerouter>(parent);
+      }
+
+      void visitStructGet(StructGet* curr) {
+        if (curr->ref->type == Type::unreachable) {
+          return;
+        }
+
+        if (curr->type.isStruct() &&
+            parent.structInfoByITableType.count(curr->type.getHeapType())) {
+          // This is a struct.get that returns an itable type;
+          // Change to return the corresponding vtable type.
+          Builder builder(*getModule());
+          replaceCurrent(builder.makeStructGet(
+            0,
+            curr->ref,
+            parent.structInfoByITableType[curr->type.getHeapType()]
+              ->javaClass.getStruct()
+              .fields[0]
+              .type));
+
+          return;
+        }
+      }
+    };
```

Comment:
The thing is that we need to do this in this specific sequence, because the children get rewritten first, if the `struct.get` on the `itable` fields get rewritten to the `struct.get` on the `vtable` field in the nesting, then there is no way to differentiate what was a `vtable` field access from an `itable` field access.